### DISCUSSION TITLE
How to build curl with websocket on windows

### DISCUSSION THREAD

**Discussion by JAMUKI** [curl/curl#11075] [Q&A] [answered]
I want to use vs2019 to compile the curl library that supports WebSockets on windows, but I don't know how to do it

**Comment by jay** [comment, marked as answer]
WebSocket support is experimental, please read about it in the curl everything book https://everything.curl.dev/libcurl/ws

curl's cmake build supports enabling websockets by using bool option `ENABLE_WEBSOCKETS=ON`.

For example you could open a [Visual Studio developer command prompt](https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170), make a separate directory to store the generated cmake build files, build curl and install:

- curl source dir `C:\stuff\curl\curl-8.0.1`
- curl cmake build dir `C:\stuff\curl\cmake-build`
- curl install dir `C:\stuff\curl\cmake-install`

and you open 'x64 Native Tools Command Prompt for VS 2019' shortcut and run

~~~batch
mkdir C:\stuff\curl\cmake-build
cd /d C:\stuff\curl\cmake-build
cmake C:\stuff\curl-8.0.1 -DCURL_USE_SCHANNEL=ON -DENABLE_WEBSOCKETS=ON -DCMAKE_INSTALL_PREFIX:PATH=C:\stuff\curl\cmake-install
cmake --build . --target INSTALL --config Release
~~~

You should see an 'Enabled protocols' of WS WSS and when you run `C:\stuff\curl\cmake-install\bin\curl -V` you should see same ws wss in protocols list.

You can use build systems other than cmake to build with Visual Studio (see projects and winbuild dirs for example) but you would have to define USE_WEBSOCKETS because there is no special build option to enable websockets. For example winbuild:

~~~
cd /d C:\stuff\curl\curl-8.0.1\winbuild
nmake /f Makefile.vc mode=dll VC=14.20 CC="cl /DUSE_WEBSOCKETS"
~~~

and curl would be output to something like C:\stuff\curl\curl-8.0.1\builds\libcurl-vc14.20-x64-release-dll-ipv6-sspi-schannel (bin,lib,include) and you copy it to your install location

**Reply by JAMUKI** [reply]
Thanks, this helped me a lot

**Comment by habi498** [comment]
This works.
For building 32-bit binaries, you should add -A Win32 in x86 Native Tools
`cmake C:\stuff\curl-8.0.1 -DCURL_USE_SCHANNEL=ON -DENABLE_WEBSOCKETS=ON -DCMAKE_INSTALL_PREFIX:PATH=C:\stuff\curl\cmake-install -A Win32`
